processor/src/chiplets/ace/tests/encoder.rs

```diff
@@ -0,0 +1,146 @@
+        let mut last_eval_node_index = circuit.instructions.len() - 1;
+        while encoded_circuit.len() < circuit_size {
+            let last_eval_node = NodeID::Eval(last_eval_node_index);
+            let square_last_instruction = Instruction {
+                node_l: last_eval_node,
+                node_r: last_eval_node,
+                op: Op::Mul,
+            };
+            let encoded_instruction =
+                Self::encode_instruction(&square_last_instruction, &layout).unwrap();
+            encoded_circuit.push(encoded_instruction);
+            last_eval_node_index += 1;
+        }
+        debug_assert_eq!(last_eval_node_index, layout.num_instructions - 1);
+
+        Ok(Self {
+            num_vars: layout.num_vars(),
+            num_eval: layout.num_instructions,
+            encoded_circuit,
+        })
+    }
+
+    // INSTRUCTION ENCODING
+    // --------------------------------------------------------------------------------------------
+
+    /// Encode an instruction as a `Felt`, packed as
+    /// `[ id_l (30 bits) || id_r (30 bits) || op (2 bits) ]`,
+    /// where `id_{l, r}` are is the index of the node in the graph, reversed
+    /// with regard to the total number of nodes.
+    pub fn encode_instruction(instruction: &Instruction, layout: &CircuitLayout) -> Option<Felt> {
+        if layout.num_nodes() > EncodedCircuit::MAX_ID as usize {
+            return None;
+        }
+
+        let id_l = layout.encoded_node_id(&instruction.node_l)?;
+        let id_r = layout.encoded_node_id(&instruction.node_r)?;
+
+        let op = match instruction.op {
+            Op::Sub => 0,
+            Op::Mul => 1,
+            Op::Add => 2,
+        };
+
+        let encoded = id_l as u64 + ((id_r as u64) << Self::ID_BITS) + (op << (2 * Self::ID_BITS));
+        Some(Felt::new(encoded))
+    }
+}
+
+impl CircuitLayout {
+    /// Same as `node_to_index`, but reverses the index relative to `num_nodes`.
+    ///
+    /// For example, the first input node has `id = layout.num_nodes() - 1` and the last
+    /// instruction produces a node with `id = 0`.
+    pub(crate) fn encoded_node_id(&self, node: &NodeID) -> Option<u32> {
+        let id = self.node_index(node)?;
+        Some((self.num_nodes() - 1 - id) as u32)
+    }
+
+    /// Returns the layout of the padded circuit ensuring the following alignment properties:
+    ///
+    /// - Number of inputs and constants are multiples of 2, ensuring the memory regions containing
+    ///   them are each word aligned, as each word contains two variables.
+    /// - The size of the circuit is double-word aligned to allow efficient un-hashing
+    /// - The number of instructions are also word-aligned.
+    fn padded(&self) -> Self {
+        // Inputs are padded to next multiple of 2 so they can be word-aligned, since each word
+        // contains two inputs.
+        // TODO(@adr1anh): does it makes sense to double-word align?
```

Comment:
Maybe not, but we can keep it for now and fix it if needed when we can run the end-to-end test